src/secrets-env.cr

```diff
@@ -4,23 +4,52 @@ require "path"
 module ENV
   DEFAULT_SECRETS_PATH = "/run/secrets"
 
-  # :nodoc:
-  ACCESSED = [] of String
+  private ACCESSED = [] of String
 
-  # Returns a compile-time generated array of each environment variable
-  # accessed by the program.
-  def self.accessed : Array(String)
-    ACCESSED
+  macro finished
+    {% if ACCESSED.empty? %}
+      private STATIC_ACCESSED = Tuple.new
+    {% else %}
+      private STATIC_ACCESSED = Tuple.new(
+        {% for key in ACCESSED %}
+          {{ key }},
+        {% end %}
+      )
+    {% end %}
   end
 
+  # Override `.[]` to enable compile-time resolution or accessed keys.
+  #
+  # Maintains the behaviour of the method of the same name.
   macro [](key)
     {{ ACCESSED << key if key.is_a? StringLiteral && !ACCESSED.includes? key }}
     ENV.fetch({{ key }})
   end
 
-  macro []?(key)
-    {{ ACCESSED << key if key.is_a? StringLiteral && !ACCESSED.includes? key }}
-    ENV.fetch({{ key }}, nil)
+  {% if compare_versions(Crystal::VERSION, "0.36.0") < 0 %}
+    {% verbatim do %}
+      # Override `.[]?` to enable compile-time resolution or accessed keys.
+      #
+      # Maintains the behaviour of the method of the same name.
+      macro []?(key)
+        {{ ACCESSED << key if key.is_a? StringLiteral && !ACCESSED.includes? key }}
+        ENV.fetch({{ key }}, nil)
+      end
+    {% end %}
+  {% else %}
+    STDERR.puts "Warning: unsupported Crystal version, ENV.accessed will only provide partial results"
```

Comment:
Should we put this log here, or set a crystal version pin in the shard.yml?
It's going to be partial but if (big if) all env accesses happen at boot and `fetch` is still overridden, it should still sort of work?

Reply:
Either way, hopefully we get a patch soon 😬

Reply:
I think this warning is good enough for this version of crystal, we can re-evaluate if `[]?` macros don't make it into the next version

Reply:
I've narrowed the impact / visibility of this in 181c9ea. Functionality should be there for _most_ cases, however if there is an attempt to do a static only lookup on 0.36.0 there's some guard rails in place.